File: algorithm/this_is_codingTest/q38.java
package this_is_codingTest;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class q38 {
    static int INF = 10001;
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int size = Integer.parseInt(st.nextToken()) + 1;
        int n = Integer.parseInt(st.nextToken());
        int result = 0;
        int[][] arr = new int[size][size];

        for(int i = 1; i < size; i++){
            for(int j = 1; j < size; j++){
                if(i == j){
                    arr[i][j] = 0;
                }
                else{
                    arr[i][j] = INF;
                }
            }
        }

        for(int i = 0; i < n; i++){
            st = new StringTokenizer(br.readLine());
            
            int a = 0;
            int b = 0;
            
            while(st.hasMoreTokens()){
                a = Integer.parseInt(st.nextToken());
                b = Integer.parseInt(st.nextToken());

                arr[a][b] = 1;
            }
        }

        for(int k = 1; k < size; k++){
            for(int i = 1; i < size; i++){
                for(int j = 1; j < size; j++){
                    arr[i][j] = Math.min(arr[i][j], arr[i][k] + arr[k][j]); //플로이드 워셜 방법사용
                }
            }
        }

        for(int i = 1; i < size; i++){
            int cnt = 0;
            for(int j = 1; j < size; j++){
                if(arr[i][j] != INF || arr[j][i] != INF){ //학생들이 연결되어있다면
                    cnt += 1;
                }
            }
            if(cnt == size - 1){ //모든 학생들의 경우의 수를 비교가능하다면
                result += 1;
            }
        }

        System.out.println(result);
    }
}

File: algorithm/this_is_codingTest/Kotlin/q38.kt
import java.util.*

private const val INFINITE = 100001

fun main() = with(System.`in`.bufferedReader()){
    var st = StringTokenizer(readLine())

    var size = st.nextToken().toInt() + 1
    var n = st.nextToken().toInt()
    var arr = Array(size){Array(size){0} }
    var result = 0

    for(i in 1 until size){
        for(j in 1 until size){
            if(i == j){
                arr[i][j] = 0
            }
            else arr[i][j] = INFINITE
        }
    }

    for(i in 0 until n){
        st = StringTokenizer(readLine())
        var a = 0
        var b = 0

        while(st.hasMoreTokens()){
            a = st.nextToken().toInt()
            b = st.nextToken().toInt()
            arr[a][b] = 1
        }
    }

    for(k in 1 until size){
        for(i in 1 until size){
            for(j in 1 until size){
                arr[i][j] = Math.min(arr[i][j], arr[i][k] + arr[k][j]) //플로이드 워셜 방법사용
            }
        }
    }

    for(i in 1 until size){
        var cnt = 0
        for(j in 1 until size){
            if(arr[i][j] != INFINITE || arr[j][i] != INFINITE){ //학생들이 연결되어있다면
                cnt += 1
            }
        }
        if(cnt == size - 1) result += 1 //모든 학생들의 경우의 수를 비교가능하다면
    }
    println(result)
}


